Add new features and examples

shujuwen2013 opened the pull request:
1. Add gradient color function to Echarts.Color
2. Add numeral string functions to Echarts.Formatter and Echarts.Utils which can be used to convert number to certain format according to some template.
3. Add date string functions to Echarts.Formatter and Echarts.Utils which can be used to convert date to certain format according to some template.
4. Some small changes to Echarts.Symbol and Utils.js
5. Add some area chart examples
   http://ec2-52-36-201-183.us-west-2.compute.amazonaws.com/purescript-echarts/new-types.html

shujuwen2013 committed:
configuration change

add "virtual-dom": "~1.3.0" in package.json

shujuwen2013 committed:
Merge pull request #1 from shujuwen2013/initial-conf-change

configuration change

shujuwen2013 committed:
add-nosymbol

a new symbol type: showing no symbol

shujuwen2013 committed:
Merge pull request #2 from shujuwen2013/add-new-types

add-nosymbol

shujuwen2013 committed:
Color in Axis Line Style

change color representation in axis line style

shujuwen2013 committed:
Add another representation for CalculableColor

Add another representation for CalculableColor: ForeignColorFunc

shujuwen2013 committed:
change the definition of AreaStyle

bringing in CalculableColor

shujuwen2013 committed:
change docs

shujuwen2013 committed:
add area-plot

add area-plot

shujuwen2013 committed:
Merge pull request #3 from shujuwen2013/add-new-types

add area-plot

shujuwen2013 committed:
refine foreign function for gradient color

refine foreign function for gradient color

shujuwen2013 committed:
Merge pull request #4 from shujuwen2013/add-new-types

refine foreign function for gradient color

shujuwen2013 committed:
add numeraljs as formatter

http://numeraljs.com/

shujuwen2013 committed:
Merge pull request #5 from shujuwen2013/add-new-types

numeral fomatter

shujuwen2013 committed:
add momentjs date-time formatter

http://momentjs.com/

shujuwen2013 committed:
Merge pull request #6 from shujuwen2013/add-new-types

Add momentjs date-time formatter

shujuwen2013 committed:
refine AreaPlot example

shujuwen2013 committed:
Add confidence band example

shujuwen2013 committed:
Merge pull request #7 from shujuwen2013/add-new-types

refine AreaPlot example

shujuwen2013 committed:
Merge pull request #8 from shujuwen2013/add-new-types

Add confidence band example

shujuwen2013 committed:
Add Percentage Area Chart

shujuwen2013 committed:
Merge pull request #9 from shujuwen2013/add-new-types

Add Percentage Area Chart

shujuwen2013 committed:
refine gradient color fucntion

shujuwen2013 committed:
change subtitle

shujuwen2013 committed:
Merge pull request #10 from shujuwen2013/add-new-types

change subtitle

shujuwen2013 committed:
remove dependency of moment.js

implemented date string converting functions in Echarts.Utils

shujuwen2013 committed:
Merge pull request #11 from shujuwen2013/add-new-types

remove dependency of moment.js

shujuwen2013 committed:
move LinearGradientInput to Echart.Color

shujuwen2013 committed:
Merge pull request #12 from shujuwen2013/add-new-types

move LinearGradientInput to Echart.Color

shujuwen2013 committed:
Add numeral string formatting functions

Add numeral string formatting functions, and remove the dependency of
numeral.js

shujuwen2013 committed:
Merge pull request #13 from shujuwen2013/add-new-types

Add numeral string formatting functions

shujuwen2013 committed:
remove dependency

shujuwen2013 committed:
Merge pull request #14 from shujuwen2013/add-new-types

remove dependency

shujuwen2013 committed:
remove the dependency of virtual-dom

shujuwen2013 committed:
Merge pull request #15 from shujuwen2013/add-new-types

remove the dependency of virtual-dom

cryogenian commented on example/PercentageArea.purs:
```diff
@@ -0,0 +1,304 @@
+
+simpleData = Value <<< Simple
+
+foreign import anotationFomatter :: forall eff. (Eff eff Unit)
```
Incorrect type

shujuwen2013 replied:
changed `forall eff. (Eff eff Unit)` to `Unit`

shujuwen2013 replied:
It turned out `anotationFomatter` here was unnecessary. So I removed it.

cryogenian commented on example/PercentageArea.purs:
```diff
@@ -0,0 +1,304 @@
+    selected: Nothing
+  }
+
+fromJustArray :: forall a. Maybe (Array a) -> Array a
```
This is `foldMap id`

cryogenian commented on example/PercentageArea.purs:
```diff
@@ -0,0 +1,304 @@
+    Nothing -> []
+    Just arr ->  arr
+
+fromJustNumber :: Maybe Number -> Number
```
I'm not sure that this is useful stuff. Could you please use just `fromMaybe zero`?

shujuwen2013 replied:
changed both `fromJustArray` and `fromJustNumber` to `fromMaybe`

cryogenian commented on example/PercentageArea.purs:
```diff
@@ -0,0 +1,304 @@
+  }
+
+
+percentageArea id = do
```
Please add type annotation

shujuwen2013 replied:
added type annotation to `percentageArea`, `areaPlot` and `confidenceBand`

cryogenian commented on src/ECharts/Formatter.purs:
```diff
@@ -15,10 +16,14 @@ import Control.Monad.Eff
-
+  | StringFormatFunc (String -> String)
+  | NumberFormatFunc (Number -> String)
+  | GenericFormatFunc GenericFormatter
```
What does `GenericFormatFunc` do?

shujuwen2013 replied:
Actually it is used to incorporate foreign javascript functions, such as:

`exports.toolTipFomatter = function(fn) {
    return  function (params,ticket,callback) {
        var res = params[0].name;
        for (var i = 0, l =  params.length; i < l; i++) {
            res += '<br/>' + params[i].seriesName + ' : ' + fn(params[i].value);        
        }   
        return res;
    }
};`

I changed the name a little bit to `ForeignFormatFunc` to let it make more senses.

cryogenian replied:
Have you tried to code this function in PureScript?

shujuwen2013 replied:
Changed  the foreign javascript function `toolTipFomatter` to purescript function

cryogenian commented on src/ECharts/Utils.purs:
```diff
@@ -1,6 +1,15 @@
 module ECharts.Utils where
 
+import Prelude
```
I suppose it would be nice if formatter functions are live in separate packages, like `Data.Number.Format` and `Data.Date.Format` or something like this. It could help with extraction these funcs to separate package later.

shujuwen2013 replied:
Yes, I think it is a good idea~

cryogenian commented on src/ECharts/Utils.purs:
```diff
@@ -11,4 +20,322 @@ i.e
+2) format the manipulated value to a string according to some template
+-}
+numeralFormatterWithValMnplt :: Number -> Number -> String -> (Number -> String)
+numeralFormatterWithValMnplt add mul template = 
```
Sorry, but I don't get idea this function looks like concept mixer. Why this function is here? Is it possible to add then multiply and then format w/o providing separate function?

shujuwen2013 replied:
One common scenario could be that: when one wants to convert a number, say 20.0, to a percentage format, say 20.0%, `numeralFormatterWithValMnplt` can help first multiply 0.01 to 20.0 (so we get 0.2), and then format 0.2 into 20.0%. 
Since the formatting and manipulation should be implemented in the same callback function, so it is here I think a good place to put the function.

cryogenian replied:
You always can use function composition like 

```
formatToPercent <<< add 12 <<< mul 0.01 
-- OR
formatToPercent <<< (_ + 12) <<< (_ * 0.01) 
```

Other point that I dislike is argument order. Since you have `Number -> Number -> String` one can't say what operation is first addition or multiplication. Function composition solves this problem. 

(OTOH you can use newtypes or records but I think it's overkill here)

shujuwen2013 replied:
@cryogenian The param of function `numeralFormatter :: String -> (Number -> String)` is a string, which represents the formate template, rather than a number value, I don't know how to manipulate the number param of the returned callback function by applying function composition. The way listed above doesn't seems to work.

cryogenian replied:
@shujuwen2013 Yeah, I see. Your example: 
- We have `Number`, e.g. 32.0
- We want adjust it and make 0.34 and then render it like "34%" 

With your function it's `numeralFormatterWithValMnplt 32 0.01 "00d%"`

I propose something like this `32 # add 2.0 >>> mul 0.01 >>> numeralFormatter "00d%" 
Although`numeralFormatter` should be rewritten in that case. 

BTW, you can omit `()` in purescript, everything is curried by default.

shujuwen2013 replied:
Thanks, `numeralFormatterWithValMnplt` was removed.

cryogenian commented on src/ECharts/Utils.purs:
```diff
@@ -11,4 +20,322 @@ i.e
+    _ -> doNothing
+  
+  where
+  converter_0 :: Number -> String
```
Could these `converter`s be generalized? I.e. be split to two function or function that takes another function?

shujuwen2013 replied:
I understand what you mean, however I still cannot think up a good way to implement that. Do you have any suggestion? Thanks @cryogenian

cryogenian replied:
First of all that converters are nested, right? like `converter_0d00a` uses `converter_0d00`. 
And these `converter_0dx0` are in fact parameterized by number of zeros. Like `converter_0d00` uses 2 zeros (`p = 2`), `converter_0d000` -- `p = 3`. I suppose you can make this `p` an argument.

shujuwen2013 replied:
changed the code to be more generalized

cryogenian commented on example/PercentageArea.purs:
```diff
@@ -0,0 +1,304 @@
+
+
+options :: Option
+options = Option $ optionDefault {
```
FYI, this code format for records is kinda deprecated (not a big deal, but you may find new indentation a bit more convenient) this would be something like 

``` purescript
options = 
  Option $ optionDefault 
    { title = Just $ Title blablabla... 
    , tooltip = Just $ Tooltip tooltipDefault 
        { trigger = Just Trigger Axis
        , textStyle = blablabla ...
        }
...

```

This allows add and remove fields w/o changing previous line :smile:

cryogenian commented on example/PercentageArea.js:
```diff
@@ -0,0 +1,7 @@
+// module PercentageArea
+
+exports.anotationFomatter = function() {
```
Typo

cryogenian replied:
And actually dead code :)

shujuwen2013 replied:
made some modifications so that PercentageArea.js was removed

cryogenian commented on example/ConfidenceBand.purs:
```diff
@@ -0,0 +1,242 @@
+      chart' <- setOption options true chart
+      pure unit
+
+data_date = ["2012-08-28 ", "2012-08-29 ", "2012-08-30 ", "2012-08-31 ", "2012-09-01 ", "2012-09-02", "2012-09-03 ", 
```
PureScript uses camelCase. This is about indentation too :) Something like 

``` purescript
dataDate = 
  [ "2012-08-28"
  , "2012-08-29" 
  ...
  ]
```

Oh, and could you please add type annotation to all type level definitions?

shujuwen2013 replied:
changed

cryogenian commented on example/ConfidenceBand.js:
```diff
@@ -0,0 +1,11 @@
+// module ConfidenceBand
+
+exports.toolTipFomatter = function(fn) {
```
This one is a bit tricky. In fact we can provide more precise type here and even use curried function. But I suppose that it's up to you to change this or not.

shujuwen2013 replied:
made some modifications so that ConfidenceBand.js was removed

cryogenian commented on src/ECharts/Utils.purs:
```diff
@@ -11,4 +20,322 @@ i.e
+
+  isZeroOrDot :: Char -> Boolean
+  isZeroOrDot c = c == '0' || c == '.'
+
```
Could you, please, remove trailing whitespaces?

shujuwen2013 replied:
removed

shujuwen2013 committed:
Merge branch 'master' into add-new-types

shujuwen2013 committed:
modify Formatter and examples

1. change genericFormatterFunc to FormatFuncNonEff
2. change foreign function: tooltipFormatter to purescript function
3. add type declarations

shujuwen2013 committed:
Merge pull request #16 from shujuwen2013/add-new-types

change Formatter and examples

shujuwen2013 committed:
small changes

remove trailing space
camelCase

shujuwen2013 committed:
Merge branch 'master' into add-new-types

cryogenian commented on src/ECharts/Utils.purs:
```diff
@@ -11,4 +20,312 @@ i.e
+    d2 = length (takeWhile isZeroOrDot numAbsStr)
+
+  converter_0d00e :: Number -> String
+  converter_0d00e num = case ((num * 1.0 + add) * mul) == 0.0 of
```
`if mul == 0.0`?

shujuwen2013 replied:
`numeralFormatterWithValMnplt` was removed

cryogenian commented on src/ECharts/Formatter.purs:
```diff
@@ -8,17 +8,19 @@ import Data.Argonaut.Core
 
 data Formatter =
   Template String
   | FormatFunc (forall eff. Array FormatParams -> Eff eff String)
```
I think you can safely remove this.

shujuwen2013 replied:
altered `FormatFunc (forall eff. Array FormatParams -> Eff eff String)` to `FormatFunc (Array FormatParams -> String)`

shujuwen2013 committed:
refine the code

generalized numeralFormatter and dateTimeFormatter functions
removed numeralFormatterWithValMnplt
altered FormatFunc (forall eff. Array FormatParams -> Eff eff String)
to FormatFunc (Array FormatParams -> String)

shujuwen2013 committed:
Merge pull request #17 from shujuwen2013/add-new-types

refine the code

cryogenian commented on src/ECharts/Utils.purs:
```diff
@@ -193,139 +162,91 @@ numeralFormatterWithValMnplt add mul template =
   where
   converter_0 :: Number -> String
-  converter_0 num = show $ Int.round ((num * 1.0 + add) * mul)
+  converter_0 = precision 0 
```
Great! Could you please rename this function to something w/o underscores like `numFormatter`, `expFormatter` etc. ?

shujuwen2013 replied:
changed

cryogenian commented on example/ConfidenceBand.purs:
```diff
@@ -161,7 +163,8 @@ options = Option $ optionDefault {
       formatter = Just $ NumberFormatFunc (
-        numeralFormatterWithValMnplt dataBase 1.0 "0.00"),
+        -- mul 1.0 to avoid some parse issue
+        (\x -> x * 1.0 + dataBase) >>> numeralFormatter "0.00e"),
```
:100:

shujuwen2013 committed:
change names

shujuwen2013 committed:
Merge pull request #18 from shujuwen2013/add-new-types

change names

shujuwen2013 committed:
Merge branch 'master' into add-new-types

shujuwen2013 committed:
Merge pull request #19 from shujuwen2013/add-new-types

Merge branch 'master' into add-new-types

shujuwen2013 committed:
Merge branch 'master' into add-new-types

shujuwen2013 committed:
update `purescript` dependency in `package.json` to `^0.8.5`

shujuwen2013 committed:
Merge branch 'master' into add-new-types

shujuwen2013 committed:
change .travis.yml

change ‘gulp concat’ into ‘travis_wait gulp concat’

shujuwen2013 committed:
Merge pull request #20 from shujuwen2013/add-new-types

change .travis.yml

cryogenian commented:
Ok, I'm going to merge and release this as `0.6.0`. Thank you very much @shujuwen2013!
